File: LeetCode/198-House-Robber(Easy)/Solution2.java
/**
 * Same with Solution1.java
 * but O(1) space
 */
class Solution {
    public int rob(int[] nums) {
        int prevNo = 0, prevYes = 0;
        
        for (int i : nums) {
            int temp = prevNo;
            prevNo = Math.max(prevNo, prevYes);
            prevYes = i + temp;
        }
        
        return Math.max(prevNo, prevYes);
    }
}

File: LeetCode/844-Backspace-String-Compare(Easy)/Solution2.cpp
/**
 * O(1) space
 */
class Solution {
public:
    bool backspaceCompare(string S, string T) {
        int i = S.size() - 1, j = T.size() - 1;
        int backspace_s = 0, backspace_t = 0;
        
        while(i >= 0 || j >= 0) {
            while (i >= 0) {
                if (S[i] == '#') {++backspace_s; --i;}
                else if (backspace_s > 0) {--backspace_s; --i;}
                else break;
            }
            
            while (j >= 0) {
                if (T[j] == '#') {++backspace_t; --j;}
                else if (backspace_t > 0) {--backspace_t; --j;}
                else break;
            }
            
            if (i >= 0 && j >= 0 && S[i] != T[j]) return false;
            
            if ((i >= 0) != (j >= 0)) return false;
            
            --i;--j;
        }
        
        return true;
    }
};